File: sdk/conformance/typescript/src/inline-uploads.test.ts
import * as assert from "node:assert/strict";
import { readFileSync } from "node:fs";
import { join } from "node:path";
import { test } from "node:test";
import { LoonFSClient } from "../../../generated/typescript/transfers.js";
import { LoonFSClient as BrowserClient } from "../../../generated/typescript-client/transfers.js";
import { TransferScope, UploadSource } from "../../../generated/typescript/transfer-runtime.js";

type Fixture = {
    name: string;
    bytes: number;
    size: number | null;
    limit?: number | null;
    feature?: boolean;
    inline?: boolean;
    error?: boolean;
    fault?: boolean;
};
const cases: Fixture[] = JSON.parse(
    readFileSync(join(__dirname, "../../../../../fixtures/inline_uploads.json"), "utf8"),
);

for (const browser of [false, true])
    for (const fixture of cases) {
        test(`inline ${browser ? "browser" : "server"} ${fixture.name}`, async () => {
            const content = Uint8Array.from({ length: fixture.bytes }, (_, i) => i % 256);
            let position = 0,
                closed = false;
            const source = {
                async *[Symbol.asyncIterator]() {
                    // Reusing this chunk catches callers retaining mutable source buffers.
                    const chunk = new Uint8Array(1024);
                    try {
                        while (position < content.length) {
                            const size = Math.min(chunk.length, content.length - position);
                            chunk.set(content.subarray(position, position + size));
                            position += size;
                            yield chunk.subarray(0, size);
                        }
                        if (fixture.fault) throw new Error("source failed at EOF");
                    } finally {
                        closed = true;
                    }
                },
            };
            const paths: string[] = [],
                commits: unknown[] = [];
            const session = {
                namespace_id: "demo",
                namespace_alias: "demo",
                upload_id: "upl_test",
                mode: "service_proxied",
            };
            const claim = {
                kind: "blob_v1",
                owner_namespace_id: "demo",
                content_id: "con_test",
                size_bytes: content.length,
                checksum: { algorithm: "sha256", value: "0".repeat(64) },
            };
            const send: typeof fetch = async (input, init) => {
                const request = new Request(input, init);
                const path = new URL(request.url).pathname;
                paths.push(path);
                let value: unknown;
                if (path.endsWith("/capabilities")) {
                    assert.equal(paths.length, 1, "publication must not reselect transport");
                    value = {
                        protocol_version: "v0",
                        api_groups: ["filesystem/v0"],
                        features: { "filesystem.commits.inline_content": fixture.feature ?? true },
                        limits:
                            fixture.limit === null ? {} : { "commit.max_inline_content_bytes": fixture.limit ?? 65536 },
                    };
                } else if (path.endsWith("/uploads")) {
                    assert.ok(!fixture.inline && !fixture.error, "inline preparation starts no upload");
                    if (fixture.feature !== false && fixture.limit !== null)
                        assert.ok(
                            position <= Math.min(fixture.limit ?? 65536, 65536) + 1024,
                            "bounded lookahead plus one caller chunk",
                        );
                    assert.equal((await request.json()).mode, "service_proxied");
                    value = { ...session, status: "open", expires_at_ms: 2000000000000 };
                } else if (path.endsWith("/content")) {
                    assert.deepEqual(
                        new Uint8Array(await request.arrayBuffer()),
                        content,
                        "fallback must preserve every byte",
                    );
                    value = { ...session, status: "open", expires_at_ms: 2000000000000, content_ref: claim };
                } else if (path.endsWith("/complete")) {
                    value = {
                        ...session,
                        status: "completed",
                        completed_at_ms: 0,
                        content_ref: claim,
                        content_token: { content_ref: claim, token: "retained" },
                    };
                } else if (path.endsWith("/commits")) {
                    const body = await request.json();
                    commits.push(body);
                    const op = body.operations[0];
                    assert.equal(body.commit_id, "stable");
                    assert.equal(body.message, "original");
                    assert.equal(op.path, "/file");
                    assert.equal(op.behavior, "replace");
                    assert.equal(op.expected_inode_id, "ino_1");
                    assert.equal(op.expected_revision_no, 1);
                    if (!browser) assert.equal(request.headers.get("Loonfs-Actor"), "writer");
                    if (fixture.inline) {
                        assert.deepEqual(new Uint8Array(Buffer.from(op.inline_content, "base64")), content);
                        assert.equal(op.content_ref, undefined);
                        assert.equal((body.content_tokens ?? []).length, 0);
                    } else {
                        assert.equal(op.inline_content, undefined);
                        assert.deepEqual(op.content_ref, claim);
                    }
                    if (commits.length === 1)
                        return Response.json({ code: "deadline_exceeded", message: "reply lost" }, { status: 503 });
                    assert.deepEqual(body, commits[0], "retry must retain the exact request");
                    value = {
                        namespace_id: "demo",
                        namespace_alias: "demo",
                        commit_id: "stable",
                        committed_seq: 1,
                        committed_at_ms: 0,
                        committed_by: "writer",
                        events: [],
                    };
                } else throw new Error(`unexpected ${path}`);
                return Response.json(value);
            };
            const options = { baseUrl: "http://api.test", token: "secret", fetch: send };
            const client = new LoonFSClient(options),
                browserClient = new BrowserClient(options);
            const requestOptions = { maxRetries: 0, headers: { "Loonfs-Actor": "writer" } };
            const prepare = browser
                ? browserClient.files.prepareStream({
                      namespace_alias: "demo",
                      content: source,
                      size_bytes: fixture.size ?? undefined,
                  })
                : client.files.prepareStream({
                      namespace_id: "demo",
                      content: source,
                      size_bytes: fixture.size ?? undefined,
                  });
            if (fixture.error) {
                await assert.rejects(prepare, (error) => !(error instanceof assert.AssertionError));
                assert.deepEqual(paths, ["/v0/capabilities"]);
            } else {
                const prepared = await prepare;
                assert.equal("inlineContent" in prepared, fixture.inline);
                const publication = {
                    path: "/file",
                    prepared,
                    commit_id: "stable",
                    message: "original",
                    behavior: "replace" as const,
                    expected_inode_id: "ino_1",
                    expected_revision_no: 1,
                };
                const publish = () =>
                    browser
                        ? browserClient.files.uploadPrepared(
                              { namespace_alias: "demo", ...publication },
                              requestOptions,
                          )
                        : client.files.uploadPrepared({ namespace_id: "demo", ...publication }, requestOptions);
                await assert.rejects(publish(), (error) => !(error instanceof assert.AssertionError));
                assert.equal((await publish()).committed_seq, 1);
                assert.equal(commits.length, 2);
            }
            assert.ok(closed, "SDK consumes or releases its source");
        });
    }

test("inline lookahead honours cancellation", async () => {
    const scope = new TransferScope({ timeoutInSeconds: 0.02 });
    const source = new UploadSource(
        {
            async *[Symbol.asyncIterator]() {
                yield new Uint8Array([1]);
                await new Promise(() => {});
            },
        },
        scope,
    );
    try {
        await assert.rejects(source.tryInline(65536));
    } finally {
        source.close();
        scope.close();
    }
});

test("inline preparation copies a caller-owned buffer", async () => {
    const content = new Uint8Array([0, 128, 255]);
    const client = new LoonFSClient({
        baseUrl: "http://api.test",
        token: "secret",
        fetch: async () =>
            Response.json({
                protocol_version: "v0",
                api_groups: [],
                features: { "filesystem.commits.inline_content": true },
                limits: { "commit.max_inline_content_bytes": 65536 },
            }),
    });
    const prepared = await client.files.prepare({ namespace_id: "demo", content });
    content.fill(1);
    assert.ok("inlineContent" in prepared);
    assert.equal(prepared.inlineContent, "AID/");
});

File: sdk/conformance/go/inline_uploads_test.go
package conformance_test

import (
	"bytes"
	"context"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/http/httptest"
	"os"
	"reflect"
	"strings"
	"sync"
	"testing"

	loonfs "github.com/loonfs/loonfs-sdk-go"
	"github.com/loonfs/loonfs-sdk-go/files"
	"github.com/loonfs/loonfs-sdk-go/option"
	"github.com/loonfs/loonfs-sdk-go/server"
)

type inlineCase struct {
	Name    string          `json:"name"`
	Bytes   int             `json:"bytes"`
	Size    *int64          `json:"size"`
	Limit   json.RawMessage `json:"limit"`
	Feature *bool           `json:"feature"`
	Inline  bool            `json:"inline"`
	Error   bool            `json:"error"`
	Fault   bool            `json:"fault"`
}

func TestInlinePreparationAndRetry(t *testing.T) {
	data, err := os.ReadFile("../fixtures/inline_uploads.json")
	if err != nil {
		t.Fatal(err)
	}
	var cases []inlineCase
	if err := json.Unmarshal(data, &cases); err != nil {
		t.Fatal(err)
	}
	for _, fixture := range cases {
		t.Run(fixture.Name, func(t *testing.T) {
			content := make([]byte, fixture.Bytes)
			for i := range content {
				content[i] = byte(i)
			}
			source := &uploadSource{reader: strings.NewReader(string(content)), fault: fixture.Fault}
			limit := int64(65536)
			limits := map[string]int64{}
			if string(fixture.Limit) != "null" {
				if len(fixture.Limit) != 0 {
					if err := json.Unmarshal(fixture.Limit, &limit); err != nil {
						t.Fatal(err)
					}
				}
				limits["commit.max_inline_content_bytes_per_operation"] = limit
			}
			feature := fixture.Feature == nil || *fixture.Feature
			session := map[string]any{"namespace_id": "demo", "upload_id": "upl_test", "mode": "service_proxied"}
			claim := map[string]any{"kind": "blob_v1", "owner_namespace_id": "demo", "content_id": "con_test", "size_bytes": len(content), "checksum": map[string]any{"algorithm": "sha256", "value": strings.Repeat("0", 64)}}
			var mu sync.Mutex
			var paths []string
			var commits []map[string]any
			endpoint := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
				mu.Lock()
				defer mu.Unlock()
				path := r.URL.Path
				paths = append(paths, path)
				w.Header().Set("Content-Type", "application/json")
				var value any
				switch {
				case strings.HasSuffix(path, "/capabilities"):
					if len(paths) != 1 {
						t.Error("publication reselected transport")
					}
					value = map[string]any{"protocol_version": "v0", "api_groups": []string{"filesystem/v0"}, "features": map[string]bool{"filesystem.commits.inline_content": feature}, "limits": limits}
				case strings.HasSuffix(path, "/uploads"):
					if fixture.Inline || fixture.Error {
						t.Error("unexpected staged upload")
					}
					if feature && string(fixture.Limit) != "null" {
						bound := limit
						if bound > 65536 {
							bound = 65536
						}
						if source.position.Load() != bound+1 {
							t.Error("lookahead did not stop at the bound")
						}
					}
					session["status"] = "open"
					session["expires_at_ms"] = 2000000000000
					value = session
				case strings.HasSuffix(path, "/content"):
					body, err := io.ReadAll(r.Body)
					if err != nil || !bytes.Equal(body, content) {
						t.Error("fallback lost prefix bytes", err)
					}
					value = session
				case strings.HasSuffix(path, "/complete"):
					session["status"] = "completed"
					session["completed_at_ms"] = 0
					session["content_ref"] = claim
					session["content_token"] = map[string]any{"content_ref": claim, "token": "retained"}
					value = session
				case strings.HasSuffix(path, "/commits"):
					var body map[string]any
					if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
						t.Error(err)
						return
					}
					commits = append(commits, body)
					op := body["operations"].([]any)[0].(map[string]any)
					if body["commit_id"] != "stable" || body["message"] != "original" || r.Header.Get("Loonfs-Actor") != "writer" || op["path"] != "/file" || op["behavior"] != "replace" || op["expected_inode_id"] != "ino_1" || op["expected_revision_no"] != float64(1) {
						t.Error("publication options changed")
					}
					if fixture.Inline {
						encoded, ok := op["inline_content"].(string)
						decoded, err := base64.StdEncoding.DecodeString(encoded)
						if !ok || err != nil || !bytes.Equal(decoded, content) || op["content_ref"] != nil {
							t.Error("wrong inline bytes")
						}
						if tokens, ok := body["content_tokens"].([]any); ok && len(tokens) != 0 {
							t.Error("unexpected inline token")
						}
					} else if op["inline_content"] != nil || op["content_ref"] == nil {
						t.Error("staged content representation changed")
					}
					if len(commits) == 1 {
						w.WriteHeader(503)
						value = map[string]any{"code": "deadline_exceeded", "message": "reply lost"}
					} else {
						if !reflect.DeepEqual(commits[0], body) {
							t.Error("retry request changed")
						}
						value = map[string]any{"namespace_id": "demo", "commit_id": "stable", "committed_seq": 1, "committed_at_ms": 0, "committed_by": "writer", "events": []any{}}
					}
				default:
					t.Error("unexpected request", path)
					w.WriteHeader(500)
				}
				json.NewEncoder(w).Encode(value)
			}))
			defer endpoint.Close()
			client := server.NewClient(option.WithBaseURL(endpoint.URL), option.WithToken("secret"))
			prepared, err := client.Files.PrepareStream(context.Background(), "demo", source, fixture.Size)
			if fixture.Error {
				if err == nil {
					t.Fatal("expected source validation failure")
				}
				mu.Lock()
				defer mu.Unlock()
				if !reflect.DeepEqual(paths, []string{"/v0/capabilities"}) {
					t.Fatal("source failure started a mutation", paths)
				}
				return
			}
			if err != nil {
				t.Fatal(err)
			}
			_, inline := prepared.(*files.InlinePreparedContent)
			if inline != fixture.Inline {
				t.Fatalf("unexpected prepared kind %T", prepared)
			}
			message, inode, revision := "original", "ino_1", int64(1)
			input := files.PreparedUploadInput{NamespaceID: "demo", Path: "/file", Prepared: prepared, CommitID: "stable", Message: &message, Behavior: loonfs.DestinationBehaviorReplace, ExpectedInodeID: &inode, ExpectedRevisionNo: &revision}
			_, err = client.Files.UploadPrepared(context.Background(), input, option.WithMaxAttempts(1), option.WithHTTPHeader(http.Header{"Loonfs-Actor": []string{"writer"}}))
			if err == nil {
				t.Fatal("expected lost-response error")
			}
			result, err := client.Files.UploadPrepared(context.Background(), input, option.WithMaxAttempts(1), option.WithHTTPHeader(http.Header{"Loonfs-Actor": []string{"writer"}}))
			if err != nil || result.CommittedSeq != 1 {
				t.Fatal("replay failed", err)
			}
			mu.Lock()
			defer mu.Unlock()
			if len(commits) != 2 {
				t.Fatal("unexpected retries", len(commits))
			}
		})
	}
}

func TestInlinePeekCancellation(t *testing.T) {
	ctx, cancel := context.WithCancel(context.Background())
	endpoint := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/v0/capabilities" {
			t.Error("mutation started during cancelled peek")
		}
		fmt.Fprint(w, `{"protocol_version":"v0","api_groups":[],"features":{"filesystem.commits.inline_content":true},"limits":{"commit.max_inline_content_bytes_per_operation":65536}}`)
	}))
	defer endpoint.Close()
	client := server.NewClient(option.WithBaseURL(endpoint.URL))
	source := &cancellingInlineSource{cancel: cancel}
	_, err := client.Files.PrepareStream(ctx, "demo", source, nil)
	if err == nil {
		t.Fatal("expected cancellation")
	}
}

type cancellingInlineSource struct{ cancel context.CancelFunc }

func (s *cancellingInlineSource) Read(p []byte) (int, error) { s.cancel(); p[0] = 1; return 1, nil }
